### PR TITLE
[WPE][GTK] API tests runner reports Crash when it shouldn't

### PR DESCRIPTION
**Pull request by dpino**
#### 90599513b257a67a43fd0dbfa8101e6ffce364cb
<pre>
[WPE][GTK] API tests runner reports Crash when it shouldn&apos;t
<a href="https://bugs.webkit.org/show_bug.cgi?id=279111">https://bugs.webkit.org/show_bug.cgi?id=279111</a>

Reviewed by NOBODY (OOPS!).

Change condition for reporting Crash on API test.

* Tools/glib/api_test_runner.py: Report Crash only if there were subtests to
run and results were empty.
(TestRunner.run_tests):
</pre>
https://github.com/WebKit/WebKit/commit/90599513b257a67a43fd0dbfa8101e6ffce364cb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65366 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44735 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17980 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69390 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15972 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67484 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52517 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16254 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52488 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11048 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68432 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41326 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56572 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33112 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37999 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13948 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14849 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/59909 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14287 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71095 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9318 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13745 "Found 1 new test failure: imported/w3c/web-platform-tests/media-source/mediasource-addsourcebuffer-mode.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59811 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9350 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56634 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60088 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/text/attributes (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7697 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1352 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/40545 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41622 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/42803 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41366 "Built successfully") | | |

### CODE REVIEW

File: Tools/glib/api_test_runner.py [hunk 1]
```diff
@@ -317,7 +317,7 @@ def run_tests(self):
                 skipped_subtests = self._test_cases_to_skip(test)
                 number_of_total_tests += len(skipped_subtests if not subtests else set(skipped_subtests).intersection(subtests))
                 results = self._run_test(test, subtests, skipped_subtests)
-                if len(results) == 0:
+                if len(subtests) > 0 and len(results) == 0:
```
Comment (csaavedra) [inline Tools/glib/api_test_runner.py:320]: I think this should be a separate change. You can land the gardening unreviewed and that makes it easier to revert this change if it causes any issue.

Comment (clopez) [reply Tools/glib/api_test_runner.py:320]: Agreed. Please split this in two commits. You can land the gardening one unreviewed.
BTW, I think there is missing a bug number on the commit log.

**Review by clopez — CHANGES_REQUESTED** [review]
Please split the patch in two

File: Tools/glib/api_test_runner.py (hunk 1 above)
Comment (dpino) [reply Tools/glib/api_test_runner.py:320]: I split the patch in two and landed the gardening part:

- https://github.com/WebKit/WebKit/pull/33106

Comment (csaavedra) [inline Tools/glib/api_test_runner.py:320]: When is `subtests <= 0`? I'm not sure I understand this change.

Comment (dpino) [reply Tools/glib/api_test_runner.py:320]: The issue that is happening in TestContextMenu when GTK4 is used is that there are not subtests to run.

https://github.com/webkit/webkit/blob/main/Tools/TestWebKitAPI/Tests/WebKitGtk/TestContextMenu.cpp#L1166

Comment (dpino) [reply Tools/glib/api_test_runner.py:320]: To test the change fixes the issue just simply run TextContextMenu test on a WebKitGTK GTK4 build. The test should pass.

Before:

```
$ Tools/Scripts/run-gtk-tests --release WebKitGTK/TestContextMenu
TEST: /app/webkit/WebKitBuild/GTK/Release/bin/TestWebKitAPI/WebKitGTK/TestContextMenu...

Unexpected crashes (1)
    /WebKitGTK/TestContextMenu
        (problem in test executable)

Ran 1 tests of 1 with 0 successful
```
Patch applied:

```bash
$ Tools/Scripts/run-gtk-tests --release WebKitGTK/TestContextMenu
TEST: /app/webkit/WebKitBuild/GTK/Release/bin/TestWebKitAPI/WebKitGTK/TestContextMenu...

Ran 1 tests of 1 with 1 successful
```

Comment (csaavedra) [reply Tools/glib/api_test_runner.py:320]: But that's because you are running only one test, right? So subtests is the subtests of that one test. Does it work if you run all the tests?

Comment (dpino) [reply Tools/glib/api_test_runner.py:320]: I doubled check and `subtests` is a list of subtests passed as an argument via '-p'. `subtests` is not the group of `subtests` within a `tests`, which is what I thought.

An alternative solution is to run TestContextMenu only for GTK3. Right now none subtests are run for GTK4.

**Comment by dpino** [comment]
Closing as this is not the right fix for this issue. I'd live the bug ticket open since it's still valid. The runner should not take a test that runs no subtests as a Crash.
